pennylane/dla/lie_closure.py

```diff
@@ -0,0 +1,282 @@
+
+    Args:
+        generators [Iterable[Union[PauliWord, PauliSentence, Operator]]]: Operators that span the vector space.
+
+    **Example**
+
+    Take the linearly dependent set of operators and span the VSpace.
+
+    .. code-block::python3
+        ops = [
+            PauliSentence({
+                PauliWord({0:"X", 1:"X"}) : 1.,
+                PauliWord({0:"Y", 1:"Y"}) : 1.
+            }),
+            PauliSentence({
+                PauliWord({0:"X", 1:"X"}) : 1.,
+            }),
+            PauliSentence({
+                PauliWord({0:"Y", 1:"Y"}) : 1.,
+            }),
+        ]
+
+        vspace = VSpace(ops)
+
+    It automatically detects that the third operator is linearly dependent on the former two, so it does not add the third operator to the basis.
+
+    >>> vspace.basis
+    [1.0 * X(0) @ X(1)
+     + 1.0 * Y(0) @ Y(1),
+     1.0 * X(0) @ X(1)]
+
+    We can also retrospectively add operators.
+
+    >>> vspace.add(PauliWord({0:"X"}))
+    [1.0 * X(0) @ X(1)
+     + 1.0 * Y(0) @ Y(1),
+     1.0 * X(0) @ X(1),
+     1.0 * X(0)]
+
+    Again, checks of linear independence are always performed. So in the following example no operator is added.
+
+    >>> vspace.add(PauliWord({0:"Y", 1:"Y"}))
+    [1.0 * X(0) @ X(1)
+     + 1.0 * Y(0) @ Y(1),
+     1.0 * X(0) @ X(1),
+     1.0 * X(0)]
+    """
+
+    def __init__(self, generators):
+
+        # Get all Pauli words that are present in at least one Pauli sentence
+        all_pws = list(reduce(set.__or__, [set(ps.keys()) for ps in generators]))
+        num_pw = len(all_pws)
+        # Create a dictionary mapping from PauliWord to row index
+        self._pw_to_idx = {pw: i for i, pw in enumerate(all_pws)}
+
+        # Initialize VSpace properties trivially
+        self._basis = []
+        rank = 0
+
+        self._M = np.zeros((num_pw, rank), dtype=float)
+        self._rank = rank
+        self._num_pw = num_pw
+
+        # Add all generators that are linearly independent
+        self.add(generators)
+
+    @property
+    def basis(self):
+        """List of basis operators of VSpace"""
+        return self._basis
+
+    def __len__(self):
+        return len(self.basis)
+
+    def add(self, other):
+        """Adding a list of PauliSentences if they are linearly independent"""
+        if isinstance(other, (PauliWord, PauliSentence, Operator)):
+            other = [other]
+
+        other = [
+            qml.pauli.pauli_sentence(op) if not isinstance(op, PauliSentence) else op
+            for op in other
+        ]
+
+        for ps in other:
+            # TODO: Potential speed-up by computing the maximal linear independent set for all current basis vectors + other, essentially algorithm1 in https://arxiv.org/abs/1012.5256
+            self._M, self._pw_to_idx, self._rank, self._num_pw, is_independent = (
+                self._add_if_independent(self._M, ps, self._pw_to_idx, self._rank, self._num_pw)
+            )
```

Comment:
Yeah good point. I have the thought of trying to avoid in-place manipulations whenever possible but cant really pin it down to something concrete (like a rule or PEP) - except for jax-compatibility

Reply:
The point is that we may want to make just checking whether a vector is linearly independent a feature of the class. In particular, in that case we would not want to update all the class attributes.

I went ahead and added that as a separate feature and named the functions a bit more sensible

Reply:
That being said, if you have suggestions how to improve that API I'd be happy to hear it as I agree its not very pretty and seems fragile

Reply:
With the new changes, I'm happy with this API.